# Introduccion al análisis de Datos.

1. Adquisicion de datos.

2. Analisis Basico  De datos.

In [1]:
# importamos la libreria a utilizar.
import pandas as pd

<h2> Leer datos </h2>
<p>
Usamos la función <code> pandas.read_csv () </code> para leer el archivo csv. En el paréntesis, colocamos la ruta del archivo junto con una comilla, para que los pandas lean el archivo en un marco de datos desde esa dirección. La ruta del archivo puede ser una URL o su dirección de archivo local. <br>
Como los datos no incluyen encabezados, podemos agregar un argumento <code> headers = None </code> dentro del método <code> read_csv () </code>, para que los pandas no establezcan automáticamente la primera fila como encabezado . <br>
También puede asignar el conjunto de datos a cualquier variable que cree.
</p>

In [2]:
other_path = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DA0101EN/auto.csv"
data = pd.read_csv(other_path, header=None)

In [4]:
#Mostramos las primeras 5 columas de data
print("Las primeras 5 columnas de data") 
data.head(5)

Las primeras 5 columnas de data


0    1            2    3    4     5            6    7      8     9   ...  \
0   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
1   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
2   1    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...   
3   2  164         audi  gas  std  four        sedan  fwd  front  99.8  ...   
4   2  164         audi  gas  std  four        sedan  4wd  front  99.4  ...   

    16    17    18    19    20   21    22  23  24     25  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  

[5 rows x 26 columns]

<h3> Agregar encabezados </h3>

En primer lugar, creamos una lista de "encabezados" que incluyen todos los nombres de columna en orden.
Luego, usamos <code> dataframe.columns = headers </code> para reemplazar los encabezados por la lista que creamos.
</p>

In [13]:
encabezados = ["simbolización", "pérdidas normalizadas", "marca", "tipo de combustible", "aspiración", "número de puertas", "estilo de carrocería",
         "ruedas motrices", "ubicación del motor", "distancia entre ejes", "longitud", "ancho", "altura", "peso en vacío", "tipo de motor",
         "número de cilindros", "tamaño del motor", "sistema de combustible", "diámetro interior", "carrera", "relación de compresión", "potencia",
         "pico-rpm", "ciudad-mpg", "carretera-mpg", "precio"]
print("Encabezados\n", encabezados)

Encabezados
 ['simbolización', 'pérdidas normalizadas', 'marca', 'tipo de combustible', 'aspiración', 'número de puertas', 'estilo de carrocería', 'ruedas motrices', 'ubicación del motor', 'distancia entre ejes', 'longitud', 'ancho', 'altura', 'peso en vacío', 'tipo de motor', 'número de cilindros', 'tamaño del motor', 'sistema de combustible', 'diámetro interior', 'carrera', 'relación de compresión', 'potencia', 'pico-rpm', 'ciudad-mpg', 'carretera-mpg', 'precio']


Reemplazamos los encabezados y volvemos a verificar nuestro marco de datos

In [15]:
data.columns = encabezados
data.head(10)

simbolización pérdidas normalizadas        marca tipo de combustible  \
0              3                     ?  alfa-romero                 gas   
1              3                     ?  alfa-romero                 gas   
2              1                     ?  alfa-romero                 gas   
3              2                   164         audi                 gas   
4              2                   164         audi                 gas   
5              2                     ?         audi                 gas   
6              1                   158         audi                 gas   
7              1                     ?         audi                 gas   
8              1                   158         audi                 gas   
9              0                     ?         audi                 gas   

  aspiración número de puertas estilo de carrocería ruedas motrices  \
0        std               two          convertible             rwd   
1        std               two          convertible             rwd   
2        std               two            hatchback             rwd   
3        std              four                sedan             fwd   
4        std              four                sedan             4wd   
5        std               two                sedan             fwd   
6        std              four                sedan             fwd   
7        std              four                wagon             fwd   
8      turbo              four                sedan             fwd   
9      turbo               two            hatchback             4wd   

  ubicación del motor  distancia entre ejes  ...  tamaño del motor  \
0               front                  88.6  ...               130   
1               front                  88.6  ...               130   
2               front                  94.5  ...               152   
3               front                  99.8  ...               109   
4               front                  99.4  ...               136   
5               front                  99.8  ...               136   
6               front                 105.8  ...               136   
7               front                 105.8  ...               136   
8               front                 105.8  ...               131   
9               front                  99.5  ...               131   

   sistema de combustible  diámetro interior  carrera relación de compresión  \
0                    mpfi               3.47     2.68                    9.0   
1                    mpfi               3.47     2.68                    9.0   
2                    mpfi               2.68     3.47                    9.0   
3                    mpfi               3.19     3.40                   10.0   
4                    mpfi               3.19     3.40                    8.0   
5                    mpfi               3.19     3.40                    8.5   
6                    mpfi               3.19     3.40                    8.5   
7                    mpfi               3.19     3.40                    8.5   
8                    mpfi               3.13     3.40                    8.3   
9                    mpfi               3.13     3.40                    7.0   

  potencia  pico-rpm ciudad-mpg carretera-mpg precio  
0      111      5000         21            27  13495  
1      111      5000         21            27  16500  
2      154      5000         19            26  16500  
3      102      5500         24            30  13950  
4      115      5500         18            22  17450  
5      110      5500         19            25  15250  
6      110      5500         19            25  17710  
7      110      5500         19            25  18920  
8      140      5500         17            20  23875  
9      160      5500         16            22      ?  

[10 rows x 26 columns]

 podemos soltar los valores faltantes a lo largo de la columna "precio" de la siguiente manera

In [16]:
data.dropna(subset=["precio"], axis=0)

simbolización pérdidas normalizadas        marca tipo de combustible  \
0                3                     ?  alfa-romero                 gas   
1                3                     ?  alfa-romero                 gas   
2                1                     ?  alfa-romero                 gas   
3                2                   164         audi                 gas   
4                2                   164         audi                 gas   
..             ...                   ...          ...                 ...   
200             -1                    95        volvo                 gas   
201             -1                    95        volvo                 gas   
202             -1                    95        volvo                 gas   
203             -1                    95        volvo              diesel   
204             -1                    95        volvo                 gas   

    aspiración número de puertas estilo de carrocería ruedas motrices  \
0          std               two          convertible             rwd   
1          std               two          convertible             rwd   
2          std               two            hatchback             rwd   
3          std              four                sedan             fwd   
4          std              four                sedan             4wd   
..         ...               ...                  ...             ...   
200        std              four                sedan             rwd   
201      turbo              four                sedan             rwd   
202        std              four                sedan             rwd   
203      turbo              four                sedan             rwd   
204      turbo              four                sedan             rwd   

    ubicación del motor  distancia entre ejes  ...  tamaño del motor  \
0                 front                  88.6  ...               130   
1                 front                  88.6  ...               130   
2                 front                  94.5  ...               152   
3                 front                  99.8  ...               109   
4                 front                  99.4  ...               136   
..                  ...                   ...  ...               ...   
200               front                 109.1  ...               141   
201               front                 109.1  ...               141   
202               front                 109.1  ...               173   
203               front                 109.1  ...               145   
204               front                 109.1  ...               141   

     sistema de combustible  diámetro interior  carrera  \
0                      mpfi               3.47     2.68   
1                      mpfi               3.47     2.68   
2                      mpfi               2.68     3.47   
3                      mpfi               3.19     3.40   
4                      mpfi               3.19     3.40   
..                      ...                ...      ...   
200                    mpfi               3.78     3.15   
201                    mpfi               3.78     3.15   
202                    mpfi               3.58     2.87   
203                     idi               3.01     3.40   
204                    mpfi               3.78     3.15   

    relación de compresión potencia  pico-rpm ciudad-mpg carretera-mpg precio  
0                      9.0      111      5000         21            27  13495  
1                      9.0      111      5000         21            27  16500  
2                      9.0      154      5000         19            26  16500  
3                     10.0      102      5500         24            30  13950  
4                      8.0      115      5500         18            22  17450  
..                     ...      ...       ...        ...           ...    ...  
200                    9.5      114      5400         23            2

<h2> Guardar conjunto de datos </h2>
<p>
En consecuencia, Pandas nos permite guardar el conjunto de datos en csv utilizando el método <code> dataframe.to_csv () </code>, puede agregar la ruta del archivo y el nombre junto con comillas entre paréntesis.
</p>
<p>
    Por ejemplo, si guarda el marco de datos <b> df </b> como <b> automobile.csv </b> en su máquina local, puede usar la sintaxis a continuación:
</p>

In [18]:
data.to_csv("automobile.csv", index=False)

También podemos leer y guardar otros formatos de archivo, podemos usar funciones similares a pd.read_csv () y df.to_csv () para otros formatos de datos, las funciones se enumeran en la siguiente tabla:

<h2>Leer / guardar otros formatos de datos</h2>



| Formato Dato  | Leer           | Guardar            |
| ------------- |:--------------:| ----------------:|
| csv           | `pd.read_csv()`  |`df.to_csv()`     |
| json          | `pd.read_json()` |`df.to_json()`    |
| excel         | `pd.read_excel()`|`df.to_excel()`   |
| hdf           | `pd.read_hdf()`  |`df.to_hdf()`     |
| sql           | `pd.read_sql()`  |`df.to_sql()`     |
| ...           |   ...          |       ...        |

<h1 id = "basic_insight"> Información básica del conjunto de datos </h1>
<p>
Después de leer los datos en el marco de datos de Pandas, es hora de que exploremos el conjunto de datos. <br>
Hay varias formas de obtener información esencial de los datos para ayudarnos a comprender mejor nuestro conjunto de datos.
</p>

<h2> Tipos de datos </h2>
<p>
Los datos tienen una variedad de tipos. <br>
Los principales tipos almacenados en los marcos de datos de Pandas son <b> objeto </b>, <b> flotante </b>, <b> int </b>, <b> bool </b> y <b> datetime64 </ b>. Para conocer mejor cada atributo, siempre es bueno para nosotros conocer el tipo de datos de cada columna. En pandas:
</p>

In [19]:
data.dtypes

simbolización               int64
pérdidas normalizadas      object
marca                      object
tipo de combustible        object
aspiración                 object
número de puertas          object
estilo de carrocería       object
ruedas motrices            object
ubicación del motor        object
distancia entre ejes      float64
longitud                  float64
ancho                     float64
altura                    float64
peso en vacío               int64
tipo de motor              object
número de cilindros        object
tamaño del motor            int64
sistema de combustible     object
diámetro interior          object
carrera                    object
relación de compresión    float64
potencia                   object
pico-rpm                   object
ciudad-mpg                  int64
carretera-mpg               int64
precio                     object
dtype: object

<h2> Describe </h2>
Si nos gustaría obtener un resumen estadístico de cada columna, como conteo, valor medio de columna, desviación estándar de columna, etc. Utilizamos el método de descripción:

In [22]:
data.describe(include = "all")

simbolización pérdidas normalizadas   marca tipo de combustible  \
count      205.000000                   205     205                 205   
unique            NaN                    52      22                   2   
top               NaN                     ?  toyota                 gas   
freq              NaN                    41      32                 185   
mean         0.834146                   NaN     NaN                 NaN   
std          1.245307                   NaN     NaN                 NaN   
min         -2.000000                   NaN     NaN                 NaN   
25%          0.000000                   NaN     NaN                 NaN   
50%          1.000000                   NaN     NaN                 NaN   
75%          2.000000                   NaN     NaN                 NaN   
max          3.000000                   NaN     NaN                 NaN   

       aspiración número de puertas estilo de carrocería ruedas motrices  \
count         205               205                  205             205   
unique          2                 3                    5               3   
top           std              four                sedan             fwd   
freq          168               114                   96             120   
mean          NaN               NaN                  NaN             NaN   
std           NaN               NaN                  NaN             NaN   
min           NaN               NaN                  NaN             NaN   
25%           NaN               NaN                  NaN             NaN   
50%           NaN               NaN                  NaN             NaN   
75%           NaN               NaN                  NaN             NaN   
max           NaN               NaN                  NaN             NaN   

       ubicación del motor  distancia entre ejes  ...  tamaño del motor  \
count                  205            205.000000  ...        205.000000   
unique                   2                   NaN  ...               NaN   
top                  front                   NaN  ...               NaN   
freq                   202                   NaN  ...               NaN   
mean                   NaN             98.756585  ...        126.907317   
std                    NaN              6.021776  ...         41.642693   
min                    NaN             86.600000  ...         61.000000   
25%                    NaN             94.500000  ...         97.000000   
50%                    NaN             97.000000  ...        120.000000   
75%                    NaN            102.400000  ...        141.000000   
max                    NaN            120.900000  ...        326.000000   

        sistema de combustible  diámetro interior  carrera  \
count                      205                205      205   
unique                       8                 39       37   
top                       mpfi               3.62     3.40   
freq                        94                 23       20   
mean                       NaN                NaN      NaN   
std                        NaN                NaN      NaN   
min                        NaN                NaN      NaN   
25%                        NaN                NaN      NaN   
50%                        NaN                NaN      NaN   
75%                        NaN                NaN      NaN   
max                        NaN                NaN      NaN   

       relación de compresión potencia  pico-rpm  ciudad-mpg carretera-mpg  \
count              205.000000      205       205  205.000000    205.000000   
unique                    NaN       60        24         NaN           NaN   
top                       NaN       68      5500         NaN           NaN   
freq                      NaN       19        37         NaN           NaN   
mean                10.142537      NaN       NaN   25.219512     30.751220   
std                  3.972040      NaN       NaN    6.542142      6.886443   
mi

<p>
Ahora, proporciona el resumen estadístico de todas las columnas, incluidos los atributos de tipo objeto. <br>
Ahora podemos ver cuántos valores únicos, que es el valor superior y la frecuencia del valor superior en las columnas de tipo objeto. <br>
Algunos valores en la tabla anterior se muestran como "NaN", esto se debe a que esos números no están disponibles con respecto a un tipo de columna en particular. <br>
</p>

<h2>Info</h2>
Otro método que puede usar para verificar su conjunto de datos es:

In [23]:
data.info

<bound method DataFrame.info of      simbolización pérdidas normalizadas        marca tipo de combustible  \
0                3                     ?  alfa-romero                 gas   
1                3                     ?  alfa-romero                 gas   
2                1                     ?  alfa-romero                 gas   
3                2                   164         audi                 gas   
4                2                   164         audi                 gas   
..             ...                   ...          ...                 ...   
200             -1                    95        volvo                 gas   
201             -1                    95        volvo                 gas   
202             -1                    95        volvo                 gas   
203             -1                    95        volvo              diesel   
204             -1                    95        volvo                 gas   

    aspiración número de puertas estilo de 

<p>
Aquí podemos ver la información de nuestro marco de datos, con las 30 filas superiores y las 30 filas inferiores.
<br> <br>
Y también nos muestra que el marco de datos completo tiene 205 filas y 26 columnas en total.
</p>